File: sdks/go/pagination.go
package hooksniff

// PaginatedIterator iterates over paginated API responses.
// The Fetch function is called with (limit, offset) and returns items + has_more flag.
type PaginatedIterator[T any] struct {
	Fetch     func(limit, offset int) ([]T, bool, error)
	Limit     int
	MaxPages  int
	offset    int
	page      int
	buf       []T
	hasMore   bool
	lastErr   error
	done      bool
}

// NewPaginatedIterator creates an iterator with default limit=20, maxPages=0 (unlimited).
func NewPaginatedIterator[T any](fetch func(limit, offset int) ([]T, bool, error)) *PaginatedIterator[T] {
	return &PaginatedIterator[T]{
		Fetch:    fetch,
		Limit:    20,
		MaxPages: 0,
		hasMore:  true,
	}
}

// Next returns the next item, or false when exhausted.
func (it *PaginatedIterator[T]) Next() (T, bool) {
	var zero T
	if it.done {
		return zero, false
	}

	// Refill buffer if empty
	for len(it.buf) == 0 {
		if !it.hasMore {
			it.done = true
			return zero, false
		}
		if it.MaxPages > 0 && it.page >= it.MaxPages {
			it.done = true
			return zero, false
		}

		items, hasMore, err := it.Fetch(it.Limit, it.offset)
		if err != nil {
			it.lastErr = err
			it.done = true
			return zero, false
		}

		it.page++
		it.offset += len(items)
		it.hasMore = hasMore
		it.buf = items

		if len(items) == 0 {
			it.done = true
			return zero, false
		}
	}

	item := it.buf[0]
	it.buf = it.buf[1:]
	return item, true
}

// Err returns the last fetch error.
func (it *PaginatedIterator[T]) Err() error {
	return it.lastErr
}

// CollectAll drains the iterator and returns all items.
func (it *PaginatedIterator[T]) CollectAll() ([]T, error) {
	var all []T
	for {
		item, ok := it.Next()
		if !ok {
			break
		}
		all = append(all, item)
	}
	return all, it.Err()
}

File: sdks/rust/src/pagination.rs
/// Pagination utilities for HookSniff SDK.
/// Provides iterator-style pagination over list endpoints.

use serde::Deserialize;

#[derive(Deserialize)]
pub struct Page<T> {
    pub data: Vec<T>,
    pub has_more: bool,
}

pub struct PaginatedIterator<T, F>
where
    F: FnMut(u32, u32) -> Result<Page<T>, Box<dyn std::error::Error>>,
{
    fetch_fn: F,
    limit: u32,
    offset: u32,
    has_more: bool,
    max_pages: u32,
    pages_fetched: u32,
}

impl<T, F> PaginatedIterator<T, F>
where
    F: FnMut(u32, u32) -> Result<Page<T>, Box<dyn std::error::Error>>,
{
    pub fn new(fetch_fn: F, limit: u32, max_pages: u32) -> Self {
        Self {
            fetch_fn,
            limit,
            offset: 0,
            has_more: true,
            max_pages,
            pages_fetched: 0,
        }
    }

    /// Fetch next page and return items
    pub fn next_page(&mut self) -> Result<Vec<T>, Box<dyn std::error::Error>> {
        if !self.has_more || self.pages_fetched >= self.max_pages {
            return Ok(Vec::new());
        }
        let page = (self.fetch_fn)(self.limit, self.offset)?;
        self.has_more = page.has_more;
        self.offset += page.data.len() as u32;
        self.pages_fetched += 1;
        Ok(page.data)
    }

    /// Collect all items into a Vec
    pub fn collect_all(&mut self) -> Result<Vec<T>, Box<dyn std::error::Error>> {
        let mut all = Vec::new();
        loop {
            let items = self.next_page()?;
            if items.is_empty() {
                break;
            }
            all.extend(items);
        }
        Ok(all)
    }
}
